kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,322 @@
+
+  else
+    log(NOTICE, "service rejected: \"", name, "\", reason: ", version.description)
+  end
+end
+
+local function negotiate_services(services_requested)
+  local services = {}
+
+  for name, versions in pairs(services_requested) do
+    if type(name) ~= "string" or type(versions) ~= "table" then
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
+
+    log_negotiation_result(name, negotiated_version)
+  end
+
+  return services
+end
+
+
+local function register_client(cluster_data_plane_purge_delay, id, client_node)
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = id, }, {
+    last_seen = time(),
+    config_hash = DECLARATIVE_EMPTY_CONFIG_HASH,
+    hostname = client_node.hostname,
+    ip = var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+  }, { ttl = cluster_data_plane_purge_delay })
+
+  if not ok then
+    log(ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return error(err)
+  end
+end
+
+local function split_services(services)
+  local accepted, accepted_n = {}, 0
+  local rejected, rejected_n = {}, 0
+  for name, version in pairs(services or empty_table) do
+    local tbl, idx
+    if version.version then
+      accepted_n = accepted_n + 1
+      tbl, idx = accepted, accepted_n
+    else
+      rejected_n = rejected_n + 1
+      tbl, idx = rejected, rejected_n
+    end
+
+    tbl[idx] = {
+      name = name,
+      version = version.version,
+      message = version.description,
+    }
+  end
+
+  return accepted, rejected
+end
+
+local function info_to_service(info)
+  return info.name, {
+    version = info.version,
+    description = info.message,
+  }
+end
+
+local function merge_servcies(accepted, rejected)
+  local services = {}
+  for _, serivce in ipairs(accepted or empty_table) do
+    local name, version = info_to_service(serivce)
+    services[name] = version
+  end
+
+  for _, serivce in ipairs(rejected or empty_table) do
+    local name, version = info_to_service(serivce)
+    services[name] = version
+  end
+
+  return services
+end
+
+function _M.init_negotiation_server(service, conf)
+  service:import("kong.services.negotiation.v1.negotiation")
+  service:set_handler("NegotiationService.NegotiateServices", function(peer, nego_req)
+    local ok, result = pcall(function()
+      local dp_id = peer.id
+      log(NOTICE, "negotiating services for DP: ", dp_id)
+      local services_requested = unwrap_services(nego_req.services_requested)
+      verify_request(nego_req)
+
+      nego_req.node.sync_status = verify_node_compatibility(nego_req.node)
+      local services = negotiate_services(services_requested)
+      register_client(conf.cluster_data_plane_purge_delay, dp_id, nego_req.node)
+
+      local accepted, rejected = split_services(services)
+
+      local nego_result = {
```

Comment:
missing CP node description

Reply:
Actually no difference but I will add it.